### PR TITLE
Algo transaction

### PR DESCRIPTION
**Pull request by JeyDev310**
added algo transaction and fixed spell

### CODE REVIEW

File: src/transaction/algo.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
+const algosdk = require('algosdk');
+const base32 = require('base32.js');
+
+export const getClient = (testnet: boolean) => {
+    if (testnet) {
+        const baseServer = 'https://testnet-algorand.api.purestake.io/ps2'
```
Comment (samuelsramko) [inline src/transaction/algo.ts:6]: make this as optional parameter - provider?: string, which defaults to TATUM_API_URL/v3/algorand/node

File: src/transaction/algo.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        const port = '';
+        const token = {'X-API-Key': 'klnvbVsn983Q3awiuWUGn2OKHkyi3U974JYtGvdt'}
+        const algodClient = new algosdk.Algodv2(token, baseServer, port);
+        return algodClient;
+    } else {
+        return null;
```
Comment (samuelsramko) [inline src/transaction/algo.ts:12]: remove the if / else, it will always consume the same data

File: src/transaction/algo.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+const algosdk = require('algosdk');
+const base32 = require('base32.js');
+
+export const getClient = (testnet: boolean) => {
+    if (testnet) {
+        const baseServer = 'https://testnet-algorand.api.purestake.io/ps2'
+        const port = '';
+        const token = {'X-API-Key': 'klnvbVsn983Q3awiuWUGn2OKHkyi3U974JYtGvdt'}
```
Comment (samuelsramko) [inline src/transaction/algo.ts:8]: make this API Key optional process.env variable, like TRON_API_KEY

File: src/transaction/algo.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        return null;
+    }
+}
+
+const waitForConfirmation = async (algodClient: any, txId: string) => {
+    let lastround = (await algodClient.status().do())['last-round'];
+    let limit = 0;
+    while (limit < 30) {
+        let pendingInfo = await algodClient.pendingTransactionInformation(txId).do();
+        if (pendingInfo['confirmed-round'] !== null && pendingInfo['confirmed-round'] > 0) {
```
Comment (samuelsramko) [inline src/transaction/algo.ts:21]: could be simplified to if (pendingInfo['confirmed-round'])

File: src/transaction/algo.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            return 1;
```
Comment (samuelsramko) [inline src/transaction/algo.ts:22]: return true / false here

File: src/transaction/algo.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            break;
+        }
+        lastround++;
+        limit++;
+        await algodClient.statusAfterBlock(lastround).do();
+    }
+    return 0;
+}
+
+export const signAlgoTransaction = async ( testnet: boolean, from: string, to: string, fee: number, amount: BigInt, privKey: string) => {
```
Comment (samuelsramko) [inline src/transaction/algo.ts:32]: missing documentation + parameter description

File: src/transaction/algo.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    let algodClient = getClient(testnet);
```
Comment (samuelsramko) [inline src/transaction/algo.ts:33]: make the parameter 1 Class - AlgoTransaction.ts in the /request/model section with validations on the fields. amount should be string with internal conversion. use fromPrivateKey for PK, also need to use PrivateKeyOrSignatureId as a base class for that - for transactions, which should be signed via Tatum KMS, where private key is not available

File: src/transaction/algo.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    let params = await algodClient.getTransactionParams().do();
+    params.fee = 1000;
```
Comment (samuelsramko) [inline src/transaction/algo.ts:35]: this is default fixed fee from Algorand?

File: src/transaction/algo.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        lastround++;
+        limit++;
+        await algodClient.statusAfterBlock(lastround).do();
+    }
+    return 0;
+}
+
+export const signAlgoTransaction = async ( testnet: boolean, from: string, to: string, fee: number, amount: BigInt, privKey: string) => {
+    let algodClient = getClient(testnet);
+    let params = await algodClient.getTransactionParams().do();
```
Comment (samuelsramko) [inline src/transaction/algo.ts:34]: use const whenever possible

File: src/transaction/algo.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        "firstRound": params.firstRound,
+        "lastRound": params.lastRound,
+        "genesisID": params.genesisID,
+        "genesisHash": params.genesisHash,
+        "note": new Uint8Array(0)
+    }
+    let signedTxn = algosdk.signTransaction(txn, secretKey);
+    let sendTx = await algodClient.sendRawTransaction(signedTxn.blob).do();
+    let confirm = await waitForConfirmation(algodClient, sendTx.txId);
+    if (confirm === 1) return sendTx.txId; else return "faild";
```
Comment (samuelsramko) [inline src/transaction/algo.ts:53]: if there is failure, throw Error()

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: src/transaction/algo.ts (hunk 6 above)
Comment (samuelsramko) [reply src/transaction/algo.ts:32]: missing unit test for that

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: src/transaction/algo.ts (hunk 8 above)
Comment (JeyDev310) [reply src/transaction/algo.ts:35]: Oh, nope. 
That was just a random value I had input for testing purposes.
Will fix it.

File: src/transaction/algo.ts (hunk 6 above)
Comment (JeyDev310) [reply src/transaction/algo.ts:32]: The unit test code is already written inside algo.spec.ts.
Should I move it to somewhere else?

File: src/transaction/algo.ts [hunk 11]
```diff
@@ -1,54 +1,51 @@
...
-        const port = '';
-        const token = {'X-API-Key': 'klnvbVsn983Q3awiuWUGn2OKHkyi3U974JYtGvdt'}
-        const algodClient = new algosdk.Algodv2(token, baseServer, port);
-        return algodClient;
-    } else {
-        return null;
-    }
+export const getClient = (testnet: boolean, provider?: string) => {
+    const baseServer = provider || `${process.env.TATUM_API_URL || TATUM_API_URL}/v3/algorand/node`;
+    const port = '';
```
Comment (samuelsramko) [inline src/transaction/algo.ts:8]: remove this, its not necessary anymore

File: src/transaction/algo.ts [hunk 12]
```diff
@@ -1,54 +1,51 @@
...
+    const token = {'X-API-Key': `${process.env.ALGO_API_KEY}`}
+    const algodClient = new algosdk.Algodv2(token, baseServer, port);
+    return algodClient;
 }
 
 const waitForConfirmation = async (algodClient: any, txId: string) => {
     let lastround = (await algodClient.status().do())['last-round'];
     let limit = 0;
     while (limit < 30) {
         let pendingInfo = await algodClient.pendingTransactionInformation(txId).do();
```
Comment (samuelsramko) [inline src/transaction/algo.ts:18]: use const

File: src/transaction/algo.ts [hunk 13]
```diff
@@ -1,54 +1,51 @@
...
     }
-    return 0;
+    return false;
 }
 
-export const signAlgoTransaction = async ( testnet: boolean, from: string, to: string, fee: number, amount: BigInt, privKey: string) => {
-    let algodClient = getClient(testnet);
+export const signAlgoTransaction = async ( testnet: boolean, tx: AlgoTransaction, provider?: string) => {
+    const algodClient = getClient(testnet, provider);
     let params = await algodClient.getTransactionParams().do();
```
Comment (samuelsramko) [inline src/transaction/algo.ts:31]: use const

File: src/transaction/algo.ts [hunk 14]
```diff
@@ -1,54 +1,51 @@
...
+    const txn = {
+        "from": tx.from,
+        "to": tx.to,
+        "fee": Number(tx.fee),
+        "amount": BigInt(tx.amount),
         "firstRound": params.firstRound,
         "lastRound": params.lastRound,
         "genesisID": params.genesisID,
         "genesisHash": params.genesisHash,
         "note": new Uint8Array(0)
```
Comment (samuelsramko) [inline src/transaction/algo.ts:45]: note should be optional parameter in the transaction in AlgoTransaction.ts

**Review by samuelsramko — CHANGES_REQUESTED** [review]

File: src/model/request/AlgoTransaction.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+    @Matches(/^[+]?((\d+(\.\d*)?)|(\.\d+))$/)
+    public fee: string;
+
+    @IsNotEmpty()
+    @IsNumberString()
+    @Matches(/^[+]?((\d+(\.\d*)?)|(\.\d+))$/)
+    public amount: string;
+
+    @IsOptional()
+    public note: string;
```
Comment (samuelsramko) [inline src/model/request/AlgoTransaction.ts:25]: Add some @MaxLength() here.

File: src/transaction/algo.ts (hunk 12 above)
Comment (samuelsramko) [reply src/transaction/algo.ts:30]: please, use const wherever possible

File: src/wallet/address.ts [hunk 16]
```diff
@@ -650,8 +650,8 @@ const convertXdcPrivateKey = (testnet: boolean, privKey: string) => {
  */
 export const generateAlgodAddressFromPrivatetKey = (privKey: string) => {
     const decoder = new base32.Decoder({type: "rfc4648"})
-    const scretKey = decoder.write(privKey).buf;
-    let mn = algosdk.secretKeyToMnemonic(scretKey)
+    const secretKey = decoder.write(privKey).buf;
+    let mn = algosdk.secretKeyToMnemonic(secretKey)
```
Comment (samuelsramko) [inline src/wallet/address.ts:654]: use const

File: src/wallet/address.ts [hunk 17]
```diff
@@ -650,8 +650,8 @@ const convertXdcPrivateKey = (testnet: boolean, privKey: string) => {
...
     const address = algosdk.mnemonicToSecretKey(mn).addr;
```
Comment (samuelsramko) [inline src/wallet/address.ts:655]: return inline variable = return algosdk.mnemo....

**Review by samuelsramko — CHANGES_REQUESTED** [review]
Also increase version in package.json by 1 in patch.

**Review by samuelsramko — APPROVED** [review]

**Pull request merged**